nameresolution/file/file.go

```diff
@@ -0,0 +1,190 @@
+package file
+
+import (
+	"crypto/rand"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"os"
+	"path/filepath"
+
+	"github.com/gofrs/flock"
+
+	"github.com/dapr/components-contrib/nameresolution"
+	"github.com/dapr/kit/config"
+	"github.com/dapr/kit/logger"
+)
+
+const (
+	dirKey           = "dir"
+	defaultDaprDir   = ".dapr"
+	defaultNamingDir = "naming"
+)
+
+type namingInfo struct {
+	AppID        string `json:"APP_ID"`
+	HostAddress  string `json:"HOST_ADDRESS"`
+	AppPort      string `json:"APP_PORT"`
+	DaprPort     string `json:"DAPR_PORT"`
+	DaprHTTPPort string `json:"DAPR_HTTP_PORT"`
+}
+
+func createNamingInfo(metadata nameresolution.Metadata) *namingInfo {
+	return &namingInfo{
+		AppID:        metadata.Properties[nameresolution.AppID],
+		HostAddress:  metadata.Properties[nameresolution.HostAddress],
+		AppPort:      metadata.Properties[nameresolution.AppPort],
+		DaprPort:     metadata.Properties[nameresolution.DaprPort],
+		DaprHTTPPort: metadata.Properties[nameresolution.DaprHTTPPort],
+	}
+}
+
+type resolver struct {
+	logger logger.Logger
+	dir    string
+}
+
+// NewResolver creates file-based name resolver.
+func NewResolver(logger logger.Logger) nameresolution.Resolver {
+	if home, err := os.UserHomeDir(); err == nil {
+		return &resolver{
+			logger: logger,
+			dir:    filepath.Join(home, defaultDaprDir, defaultNamingDir),
+		}
+	}
+
+	return &resolver{
+		logger: logger,
+	}
+}
+
+// Init registers service for file-based name resolver.
+func (r *resolver) Init(metadata nameresolution.Metadata) error {
+	if err := r.prepareResolverDir(metadata); err != nil {
+		return err
+	}
+
+	info := createNamingInfo(metadata)
+
+	fileLock := r.newFileLock(info.AppID)
+	if err := fileLock.Lock(); err != nil {
+		r.logger.Errorf("fail to lock: %s", fileLock.Path())
+		return err
+	}
+
+	defer func(lck *flock.Flock) {
+		if err := lck.Unlock(); err != nil {
+			r.logger.Errorf("fail to unlock: %s, with error: %s", lck.Path(), err)
+		}
+	}(fileLock)
+
+	f := filepath.Join(r.dir, info.AppID)
+	namingInfos, err := loadNamingInfo(f)
+	if err != nil {
+		r.logger.Errorf("fail to load naming info from file: %s", f)
+		return err
+	}
+
+	index := -1
+	for i, ni := range namingInfos {
+		if ni.AppID == info.AppID &&
```

Comment:
How about converting check into a function, then any changes can happen in that?

Reply:
refactor it into separated method, pls. check.